File: CashuWallet/Core/CashuRequestListener.swift
import Foundation
import SwiftUI

/// An incoming NUT-18 payment held for an explicit user decision on the
/// receive screen — because its mint isn't tracked yet (claiming would
/// silently add it) or because auto-claim is disabled.
struct PendingCashuClaimApproval: Identifiable, Equatable {
    /// Gift-wrap event id; doubles as the processed-ids key.
    let id: String
    let tokenString: String
    let requestId: String?
    let mintUrl: String
    /// Sum of the payload's proof amounts, for display in the prompt.
    let amount: UInt64
    let memo: String?
}

/// NUT-18 receive-side listener. Foreground-only: opens a NIP-17 relay subscription
/// at app launch, decrypts gift wraps, parses PaymentRequestPayload from the inner
/// rumor, and forwards to the auto-claim path in WalletManager.
@MainActor
final class CashuRequestListener: ObservableObject {
    static let shared = CashuRequestListener()

    @Published private(set) var isRunning: Bool = false

    /// Payments waiting for the user's claim/decline decision (unknown mint,
    /// or auto-claim disabled).
    @Published private(set) var pendingApprovals: [PendingCashuClaimApproval] = []

    private var client: NostrInboxClient?
    private weak var walletManager: WalletManager?

    // Gift wraps are fetched over a generous fixed lookback window. NIP-59
    // backdates each gift wrap's `created_at` up to ~2 days, so a tight or
    // forward-advancing `since` floor silently drops later payments. We instead
    // re-scan a wide window every start and prevent re-processing by remembering
    // the gift-wrap event ids we've already handled.
    private let lookbackWindow: TimeInterval = 7 * 24 * 60 * 60
    private let processedIdsKey = StorageKeys.cashuRequestsProcessedNIP17Ids
    private let maxProcessedIds = 1000
    private var processedIds: Set<String> = []
    private var processedOrder: [String] = []

    private init() {}

    func attach(walletManager: WalletManager) {
        self.walletManager = walletManager
    }

    func start() async {
        guard !isRunning else { return }
        guard SettingsManager.shared.enablePaymentRequests else {
            AppLogger.wallet.notice("CashuRequestListener: payment requests disabled in settings — not starting")
            return
        }
        let nostr = NostrService.shared
        guard nostr.isInitialized,
              !nostr.publicKeyHex.isEmpty,
              let privHex = nostr.getPrivateKeyHex(),
              let privateKey = Data(hex: privHex) else {
            AppLogger.wallet.error("CashuRequestListener: NostrService not initialized")
            return
        }
        let relays = SettingsManager.shared.nostrRelays
        guard !relays.isEmpty else {
            AppLogger.wallet.error("CashuRequestListener: no Nostr relays configured — cannot receive Cashu Request payments")
            return
        }
        loadProcessedIds()
        let since = Int64(Date().timeIntervalSince1970 - lookbackWindow)

        let pubkeyHex = nostr.publicKeyHex
        let client = NostrInboxClient(
            pubkeyHex: pubkeyHex,
            relays: relays,
            since: since
        ) { [weak self] event in
            await self?.handle(event: event, recipientPrivateKey: privateKey)
        }
        self.client = client
        await client.start()
        isRunning = true
        AppLogger.wallet.notice("CashuRequestListener: started on \(relays.count) relays, pubkey=\(String(pubkeyHex.prefix(8)), privacy: .public), since=\(since)")
    }

    func stop() async {
        guard let client else { return }
        await client.stop()
        self.client = nil
        isRunning = false
    }

    /// Forget processed gift-wrap ids at a wallet boundary. The ids belong to
    /// the previous wallet's Nostr inbox; a new wallet has a new keypair, and a
    /// re-restored seed should re-attempt claims rather than skip them.
    func resetForWalletBoundary() {
        processedIds = []
        processedOrder = []
        pendingApprovals = []
        UserDefaults.standard.removeObject(forKey: processedIdsKey)
    }

    // MARK: - Event handling

    private func handle(event: NostrIncomingEvent, recipientPrivateKey: Data) async {
        guard event.kind == 1059 else { return }
        guard !processedIds.contains(event.id) else { return }
        AppLogger.wallet.notice("CashuRequestListener: gift wrap received id=\(String(event.id.prefix(8)), privacy: .public) createdAt=\(event.createdAt)")

        let rumor: NostrRumor
        do {
            rumor = try NIP17.unwrap(giftWrap: event, recipientPrivateKey: recipientPrivateKey)
        } catch {
            // Not encrypted for us (or an unrelated DM) — it can never succeed,
            // so mark it handled and stop reconsidering it.
            AppLogger.wallet.notice("CashuRequestListener: NIP-17 unwrap failed for \(String(event.id.prefix(8)), privacy: .public): \(String(describing: error), privacy: .public)")
            markProcessed(event.id)
            return
        }
        guard rumor.kind == 14 else {
            markProcessed(event.id)
            return
        }
        switch await tryClaim(rumorContent: rumor.content, eventId: event.id) {
        case .claimed, .unclaimable:
            markProcessed(event.id)
        case .transientFailure, .awaitingApproval:
            break  // leave unmarked so a later run retries / re-offers
        }
    }

    private enum ClaimOutcome {
        case claimed            // redeemed successfully
        case unclaimable        // malformed / un-redeemable payload — never retry
        case transientFailure   // redeem failed (mint/network) — retry later
        case awaitingApproval   // unknown mint — queued for explicit user approval
    }

    private func tryClaim(rumorContent content: String, eventId: String) async -> ClaimOutcome {
        guard let data = content.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return .unclaimable
        }
        // NUT-18 PaymentRequestPayload:
        // { "id": "<request_uuid>", "memo": "...", "mint": "<url>", "unit": "sat", "proofs": [ ... ] }
        let requestId = json["id"] as? String
        guard let mintUrl = json["mint"] as? String,
              let proofs = json["proofs"] as? [[String: Any]] else {
            AppLogger.wallet.notice("CashuRequestListener: malformed PaymentRequestPayload")
            return .unclaimable
        }
        let unit = (json["unit"] as? String) ?? "sat"
        let memo = json["memo"] as? String
        guard let tokenString = buildV3Token(mint: mintUrl, proofs: proofs, unit: unit, memo: memo) else {
            AppLogger.wallet.notice("CashuRequestListener: could not build token from payload")
            return .unclaimable
        }
        guard let walletManager else {
            AppLogger.wallet.error("CashuRequestListener: walletManager not attached")
            return .transientFailure
        }

        // Silent claiming needs both: auto-claim enabled, and a mint the user
        // already trusts (claiming creates a CDK wallet for the mint and adds
        // it to the tracked list — never do that without consent). Everything
        // else queues for an explicit decision on the receive screen.
        let mintKnown = walletManager.isMintKnown(url: mintUrl)
        let autoClaim = SettingsManager.shared.receivePaymentRequestsAutomatically
        guard autoClaim && mintKnown else {
            enqueueApproval(
                PendingCashuClaimApproval(
                    id: eventId,
                    tokenString: tokenString,
                    requestId: requestId,
                    mintUrl: mintUrl,
                    amount: proofsTotalAmount(proofs),
                    memo: memo
                ),
                reason: mintKnown ? "auto-claim off" : "unknown mint"
            )
            return .awaitingApproval
        }

        return await claimNow(tokenString: tokenString, requestId: requestId)
    }

    private func claimNow(tokenString: String, requestId: String?) async -> ClaimOutcome {
        guard let walletManager else {
            AppLogger.wallet.error("CashuRequestListener: walletManager not attached")
            return .transientFailure
        }
        do {
            // A gift wrap can arrive exactly as the app backgrounds; hold a background-task
            // assertion so this SQLite-writing redeem finishes before suspension.
            let amount = try await withBackgroundWriteAssertion("cashu-request-claim") {
                try await walletManager.receiveCashuRequestPayment(
                    tokenString: tokenString,
                    requestId: requestId
                )
            }
            AppLogger.wallet.notice("CashuRequestListener: claimed \(amount) sat for request \(requestId ?? "—", privacy: .public)")
            return .claimed
        } catch {
            AppLogger.wallet.error("CashuRequestListener: redeem failed (will retry): \(String(describing: error), privacy: .public)")
            return .transientFailure
        }
    }

    // MARK: - Unknown-mint approval

    /// Cap so a spammer pushing payments from throwaway mints can't grow the
    /// queue without bound. Overflow events stay unprocessed and are re-offered
    /// on a later scan once the queue drains. Generous because with auto-claim
    /// off every legitimate payment queues here too.
    private static let maxPendingApprovals = 50

    private func enqueueApproval(_ approval: PendingCashuClaimApproval, reason: String) {
        guard !pendingApprovals.contains(where: { $0.id == approval.id }) else { return }
        guard pendingApprovals.count < Self.maxPendingApprovals else {
            AppLogger.wallet.notice("CashuRequestListener: approval queue full — deferring \(String(approval.id.prefix(8)), privacy: .public)")
            return
        }
        AppLogger.wallet.notice("CashuRequestListener: payment from \(approval.mintUrl, privacy: .public) queued for approval (\(reason, privacy: .public))")
        pendingApprovals.append(approval)
    }

    /// User accepted from the receive screen: claim the payment (this adds the
    /// mint to the wallet), then auto-claim any other queued payments from
    /// mints that are now known. Throws so the screen's failure state handles
    /// retry; the approval stays queued and unprocessed until a claim succeeds
    /// or the user declines.
    func claimApproved(_ approval: PendingCashuClaimApproval) async throws -> UInt64 {
        guard let walletManager else {
            throw WalletError.notInitialized
        }
        let amount = try await withBackgroundWriteAssertion("cashu-request-claim") {
            try await walletManager.receiveCashuRequestPayment(
                tokenString: approval.tokenString,
                requestId: approval.requestId
            )
        }
        pendingApprovals.removeAll { $0.id == approval.id }
        markProcessed(approval.id)
        AppLogger.wallet.notice("CashuRequestListener: user approved claim of \(amount) sat from \(approval.mintUrl, privacy: .public)")
        await claimEligibleQueuedApprovals()
        return amount
    }

    /// User declined: drop the payment and never ask about this event again.
    func decline(_ approval: PendingCashuClaimApproval) {
        pendingApprovals.removeAll { $0.id == approval.id }
        markProcessed(approval.id)
        AppLogger.wallet.notice("CashuRequestListener: user declined payment from \(approval.mintUrl, privacy: .public)")
    }

    /// Drain queued approvals that no longer need a decision: auto-claim is on
    /// and the mint is known (the user just approved a payment from that mint,
    /// added the mint manually, or re-enabled auto-claim). No-op in manual
    /// mode — there every payment gets its own confirmation.
    func claimEligibleQueuedApprovals() async {
        guard SettingsManager.shared.receivePaymentRequestsAutomatically else { return }
        guard let walletManager else { return }
        for approval in pendingApprovals where walletManager.isMintKnown(url: approval.mintUrl) {
            pendingApprovals.removeAll { $0.id == approval.id }
            switch await claimNow(tokenString: approval.tokenString, requestId: approval.requestId) {
            case .claimed, .unclaimable:
                markProcessed(approval.id)
            case .transientFailure, .awaitingApproval:
                break
            }
        }
    }

    private func proofsTotalAmount(_ proofs: [[String: Any]]) -> UInt64 {
        proofs.reduce(UInt64(0)) { total, proof in
            let amount = (proof["amount"] as? NSNumber)?.uint64Value ?? 0
            return total &+ amount
        }
    }

    // MARK: - De-duplication

    private func loadProcessedIds() {
        let stored = UserDefaults.standard.stringArray(forKey: processedIdsKey) ?? []
        processedOrder = stored
        processedIds = Set(stored)
    }

    private func markProcessed(_ id: String) {
        guard processedIds.insert(id).inserted else { return }
        processedOrder.append(id)
        if processedOrder.count > maxProcessedIds {
            let overflow = processedOrder.count - maxProcessedIds
            for removed in processedOrder.prefix(overflow) { processedIds.remove(removed) }
            processedOrder.removeFirst(overflow)
        }
        UserDefaults.standard.set(processedOrder, forKey: processedIdsKey)
    }

    /// Build a NUT-00 V3 cashu token string from a mint + proofs payload.
    /// Format: `cashuA` + base64url(no padding)(JSON({token:[{mint, proofs}], unit, memo})).
    private func buildV3Token(mint: String, proofs: [[String: Any]], unit: String?, memo: String?) -> String? {
        var token: [String: Any] = ["token": [["mint": mint, "proofs": proofs]]]
        if let unit { token["unit"] = unit }
        if let memo { token["memo"] = memo }
        guard let data = try? JSONSerialization.data(withJSONObject: token) else { return nil }
        return "cashuA" + Base64URL.encode(data)
    }
}

File: android/app/src/main/java/org/cashu/wallet/Core/CashuRequestListener.kt
package org.cashu.wallet.Core

import java.util.Base64
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.longOrNull
import org.cashu.wallet.Core.Wallet.userFacingWalletMessage
import org.cashu.wallet.Models.PendingReceiveToken

data class CashuRequestListenerState(
    val isRunning: Boolean = false,
    val lastError: String? = null,
    /** One-shot prompt only; the durable payment lives in pending-receive storage. */
    val heldForApproval: PendingReceiveToken? = null,
)

internal enum class CashuRequestClaimAction {
    ClaimSilently,
    HoldForApproval,
}

internal object CashuRequestApprovalPolicy {
    fun action(autoClaim: Boolean, mintKnown: Boolean): CashuRequestClaimAction =
        if (autoClaim && mintKnown) {
            CashuRequestClaimAction.ClaimSilently
        } else {
            CashuRequestClaimAction.HoldForApproval
        }
}

/**
 * Foreground NUT-18 receive listener.
 *
 * Matches iOS trust semantics: a payment is only claimed without interaction
 * when automatic claiming is enabled and its mint is already tracked. Every
 * other valid payment is stored as a pending receive before its relay event is
 * acknowledged, so it survives restarts and remains claimable from History.
 */
class CashuRequestListener(
    private val nostrService: NostrService,
    private val settingsManager: SettingsManager,
    private val walletManager: WalletManager,
    private val walletStore: WalletStore,
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private var client: NostrInboxClient? = null
    private val eventsInFlight = ConcurrentHashMap.newKeySet<String>()
    @Volatile
    private var walletGeneration = 0L
    private val mutableState = MutableStateFlow(CashuRequestListenerState())
    val state: StateFlow<CashuRequestListenerState> = mutableState.asStateFlow()

    fun start() {
        if (!settingsManager.state.value.enablePaymentRequests) {
            stop()
            return
        }
        if (client != null) return
        val nostr = nostrService.state.value
        val privateKeyHex = nostrService.currentPrivateKey()
        if (!nostr.isInitialized || nostr.publicKeyHex.isBlank() || privateKeyHex.isNullOrBlank()) {
            mutableState.value = mutableState.value.copy(lastError = "Nostr key is not initialized.")
            return
        }
        val relays = settingsManager.state.value.nostrRelays
            .map(String::trim)
            .filter { it.startsWith("ws://") || it.startsWith("wss://") }
            .distinct()
        if (relays.isEmpty()) {
            mutableState.value = mutableState.value.copy(lastError = "No Nostr relays configured.")
            return
        }

        // NIP-59 gift wraps can be backdated by roughly two days. Re-scan a
        // fixed wide window on each start and de-duplicate by event id instead
        // of advancing a `since` cursor that could skip a later backdated wrap.
        val since = (System.currentTimeMillis() / 1000) - LookbackWindowSeconds
        val recipientPrivateKey = NIP44.hexToBytes(privateKeyHex)
        val generation = walletGeneration
        client = NostrInboxClient(
            pubkeyHex = nostr.publicKeyHex,
            relays = relays,
            since = since,
        ) { event ->
            handle(event, recipientPrivateKey, generation)
        }.also { it.start() }
        mutableState.value = mutableState.value.copy(isRunning = true, lastError = null)
        AppLogger.wallet.info("CashuRequestListener: started on ${relays.size} relays")
    }

    fun stop() {
        client?.stop()
        client = null
        mutableState.value = mutableState.value.copy(isRunning = false)
    }

    /** Invalidate callbacks before the wallet repository starts changing. */
    fun pauseForWalletBoundary() {
        walletGeneration += 1
        stop()
        eventsInFlight.clear()
    }

    /** Clear identity-bound listener state after create/restore/delete. */
    fun resetForWalletBoundary() {
        pauseForWalletBoundary()
        walletStore.saveProcessedNIP17EventIds(emptyList())
        mutableState.value = CashuRequestListenerState()
    }

    private suspend fun handle(
        event: NostrIncomingEvent,
        recipientPrivateKey: ByteArray,
        generation: Long,
    ) {
        if (event.kind != 1059 || isProcessed(event.id) || !eventsInFlight.add(event.id)) return
        try {
            val rumor = runCatching { NIP17.unwrap(event, recipientPrivateKey) }
                .onFailure { AppLogger.wallet.debug("CashuRequestListener: NIP-17 unwrap failed") }
                .getOrNull()
            if (rumor == null || rumor.kind != 14) {
                if (generation == walletGeneration) markProcessed(event.id)
                return
            }
            if (generation != walletGeneration) return

            when (tryClaim(rumor.content)) {
                ClaimOutcome.Claimed,
                ClaimOutcome.Held,
                ClaimOutcome.Unclaimable -> if (generation == walletGeneration) markProcessed(event.id)
                ClaimOutcome.TransientFailure -> Unit
            }
        } finally {
            eventsInFlight.remove(event.id)
        }
    }

    private enum class ClaimOutcome {
        Claimed,
        Held,
        Unclaimable,
        TransientFailure,
    }

    private suspend fun tryClaim(content: String): ClaimOutcome {
        val payload = runCatching { paymentPayloadToToken(content) }
            .onFailure { AppLogger.wallet.debug("CashuRequestListener: malformed payment payload") }
            .getOrNull() ?: return ClaimOutcome.Unclaimable

        return when (
            CashuRequestApprovalPolicy.action(
                autoClaim = settingsManager.state.value.receivePaymentRequestsAutomatically,
                mintKnown = walletManager.isMintKnown(payload.mintUrl),
            )
        ) {
            CashuRequestClaimAction.ClaimSilently -> claimNow(payload)
            CashuRequestClaimAction.HoldForApproval -> holdForApproval(payload)
        }
    }

    private suspend fun claimNow(payload: PaymentPayloadToken): ClaimOutcome = try {
        walletManager.receiveCashuRequestPayment(
            tokenString = payload.token,
            requestId = payload.requestId,
        )
        ClaimOutcome.Claimed
    } catch (cancelled: CancellationException) {
        throw cancelled
    } catch (error: Throwable) {
        AppLogger.wallet.error("CashuRequestListener: redeem failed; will retry", error)
        mutableState.value = mutableState.value.copy(lastError = error.userFacingWalletMessage)
        ClaimOutcome.TransientFailure
    }

    private suspend fun holdForApproval(payload: PaymentPayloadToken): ClaimOutcome {
        val pending = walletManager.state.value.pendingReceiveTokens
        if (pending.any { it.token == payload.token }) return ClaimOutcome.Held
        if (pending.count { it.cashuRequestId != null } >= MaxHeldPayments) {
            AppLogger.wallet.info("CashuRequestListener: held-payment backlog full")
            return ClaimOutcome.TransientFailure
        }

        val held = PendingReceiveToken(
            tokenId = UUID.randomUUID().toString(),
            token = payload.token,
            amount = payload.amount,
            dateEpochMillis = System.currentTimeMillis(),
            mintUrl = payload.mintUrl,
            unit = payload.unit,
            cashuRequestId = payload.requestId ?: "",
            memo = payload.memo,
        )
        walletManager.savePendingReceiveToken(held)
        walletManager.loadTransactions()
        mutableState.value = mutableState.value.copy(heldForApproval = held, lastError = null)
        return ClaimOutcome.Held
    }

    suspend fun claimHeldPayment(pending: PendingReceiveToken): Long {
        val amount = walletManager.claimPendingReceiveToken(pending)
        if (mutableState.value.heldForApproval?.tokenId == pending.tokenId) {
            mutableState.value = mutableState.value.copy(heldForApproval = null)
        }
        claimEligibleHeldPayments()
        return amount
    }

    fun declineHeldPayment(pending: PendingReceiveToken) {
        walletManager.removePendingReceiveToken(pending.tokenId)
        if (mutableState.value.heldForApproval?.tokenId == pending.tokenId) {
            mutableState.value = mutableState.value.copy(heldForApproval = null)
        }
        scope.launch { walletManager.loadTransactions() }
    }

    /** Hide only the one-shot prompt; the durable History item remains. */
    fun dismissHeldPayment() {
        mutableState.value = mutableState.value.copy(heldForApproval = null)
    }

    /** Claim known-mint payments that became eligible after auto-claim was enabled. */
    suspend fun claimEligibleHeldPayments() {
        if (!settingsManager.state.value.receivePaymentRequestsAutomatically) return
        val eligible = walletManager.state.value.pendingReceiveTokens.filter {
            it.cashuRequestId != null && walletManager.isMintKnown(it.mintUrl)
        }
        eligible.forEach { pending ->
            try {
                walletManager.claimPendingReceiveToken(pending)
                if (mutableState.value.heldForApproval?.tokenId == pending.tokenId) {
                    mutableState.value = mutableState.value.copy(heldForApproval = null)
                }
            } catch (cancelled: CancellationException) {
                throw cancelled
            } catch (error: Throwable) {
                AppLogger.wallet.error("CashuRequestListener: held payment remains claimable", error)
            }
        }
    }

    private fun isProcessed(id: String): Boolean = id in walletStore.loadProcessedNIP17EventIds()

    private fun markProcessed(id: String) {
        val current = walletStore.loadProcessedNIP17EventIds()
        if (id in current) return
        walletStore.saveProcessedNIP17EventIds((current + id).takeLast(MaxProcessedEventIds))
    }

    data class PaymentPayloadToken(
        val token: String,
        val requestId: String?,
        val mintUrl: String,
        val amount: Long,
        val unit: String,
        val memo: String?,
    )

    companion object {
        private const val LookbackWindowSeconds = 7L * 24 * 60 * 60
        private const val MaxProcessedEventIds = 1_000
        private const val MaxHeldPayments = 50
        private val payloadJson = Json { ignoreUnknownKeys = true; encodeDefaults = true }

        fun paymentPayloadToToken(content: String): PaymentPayloadToken {
            val fields = payloadJson.parseToJsonElement(content).jsonObject
            val mintUrl = fields["mint"]?.jsonPrimitive?.contentOrNull
                ?: throw IllegalArgumentException("Payment payload mint missing.")
            val proofs = fields["proofs"]?.jsonArray
                ?: throw IllegalArgumentException("Payment payload proofs missing.")
            val unit = fields["unit"]?.jsonPrimitive?.contentOrNull ?: "sat"
            val memo = fields["memo"]?.jsonPrimitive?.contentOrNull
            val token = JsonObject(
                buildMap {
                    put(
                        "token",
                        JsonArray(
                            listOf(
                                JsonObject(
                                    mapOf(
                                        "mint" to JsonPrimitive(mintUrl),
                                        "proofs" to proofs,
                                    ),
                                ),
                            ),
                        ),
                    )
                    put("unit", JsonPrimitive(unit))
                    if (!memo.isNullOrBlank()) put("memo", JsonPrimitive(memo))
                },
            )
            val encoded = Base64.getUrlEncoder()
                .withoutPadding()
                .encodeToString(payloadJson.encodeToString(token).toByteArray(Charsets.UTF_8))
            return PaymentPayloadToken(
                token = "cashuA$encoded",
                requestId = fields["id"]?.jsonPrimitive?.contentOrNull,
                mintUrl = mintUrl,
                amount = proofs.sumOf { proof ->
                    proof.jsonObject["amount"]?.jsonPrimitive?.longOrNull?.coerceAtLeast(0) ?: 0L
                },
                unit = unit,
                memo = memo,
            )
        }
    }
}
